File: internal/plugins/react/rules/jsx_props_no_multi_spaces/jsx_props_no_multi_spaces_test.go
package jsx_props_no_multi_spaces

import (
	"testing"

	"github.com/web-infra-dev/rslint/internal/plugins/react/rules/fixtures"
	"github.com/web-infra-dev/rslint/internal/rule_tester"
)

func TestJsxPropsNoMultiSpacesRule(t *testing.T) {
	rule_tester.RunRuleTester(fixtures.GetRootDir(), "tsconfig.json", t, &JsxPropsNoMultiSpacesRule, []rule_tester.ValidTestCase{
		{
			Code: `var x = <div foo="bar" baz="qux" />`,
			Tsx:  true,
		},
		{
			Code: `var x = <div foo="bar" />`,
			Tsx:  true,
		},
		{
			Code: `var x = <div
  foo="bar"
  baz="qux"
/>`,
			Tsx: true,
		},
		{
			Code: `var x = <div />`,
			Tsx:  true,
		},
		{
			// Namespaced attribute: single space is valid
			Code: `var x = <div foo:bar="baz" qux="quux" />`,
			Tsx:  true,
		},
		{
			// Spread attribute: single space is valid
			Code: `var x = <div {...props} foo="bar" />`,
			Tsx:  true,
		},
		{
			// No props at all
			Code: `var x = <App />`,
			Tsx:  true,
		},
		{
			// Single prop
			Code: `var x = <App foo />`,
			Tsx:  true,
		},
		{
			// Compound component name with single space
			Code: `var x = <Foo.Bar baz="quux" />`,
			Tsx:  true,
		},
		{
			// Long compound component name
			Code: `var x = <Foo.Bar.Baz.Qux quux="corge" />`,
			Tsx:  true,
		},
		{
			// Spaces inside attribute value are OK
			Code: `var x = <App foo="with  spaces   " bar />`,
			Tsx:  true,
		},
		{
			// Comment between props bridges the gap (no blank line)
			Code: `var x = <button
  title="Some button"
  // this is a comment
  onClick={() => {}}
  type="button"
/>`,
			Tsx: true,
		},
		{
			// Two comments between props bridge the gap
			Code: `var x = <button
  title="Some button"
  // first comment
  // second comment
  onClick={() => {}}
/>`,
			Tsx: true,
		},
		{
			// Inline comment + line comment
			Code: `var x = <App
  foo="bar" // comment
  // comment
  bar=""
/>`,
			Tsx: true,
		},
		{
			// TypeScript generic tag name: single space
			Code: `var x = <App<T> foo bar />`,
			Tsx: true,
		},
		{
			// Multi-line prop value: next prop starts on end line of previous prop — treated as same line
			Code: `var x = <App foo={{
  a: 1
}} bar />`,
			Tsx: true,
		},
		{
			// TypeScript generic with proper spacing
			Code: `var x = <App<T> foo="bar" baz="qux" />`,
			Tsx:  true,
		},
	}, []rule_tester.InvalidTestCase{
		{
			Code:   `var x = <div foo="bar"  baz="qux" />`,
			Tsx:    true,
			Output: []string{`var x = <div foo="bar" baz="qux" />`},
			Errors: []rule_tester.InvalidTestCaseError{
				{
					MessageId: "onlyOneSpace",
					Line:      1,
					Column:    25,
				},
			},
		},
		{
			Code:   `var x = <div  foo="bar" />`,
			Tsx:    true,
			Output: []string{`var x = <div foo="bar" />`},
			Errors: []rule_tester.InvalidTestCaseError{
				{
					MessageId: "onlyOneSpace",
					Line:      1,
					Column:    15,
				},
			},
		},
		{
			Code: `var x = <div
  foo="bar"

  baz="qux"
/>`,
			Tsx: true,
			Errors: []rule_tester.InvalidTestCaseError{
				{
					MessageId: "noLineGap",
					Line:      4,
					Column:    3,
				},
			},
		},
		{
			// Namespaced attribute: multiple spaces should be reported
			Code:   `var x = <div foo:bar="baz"  qux="quux" />`,
			Tsx:    true,
			Output: []string{`var x = <div foo:bar="baz" qux="quux" />`},
			Errors: []rule_tester.InvalidTestCaseError{
				{
					MessageId: "onlyOneSpace",
					Line:      1,
					Column:    29,
				},
			},
		},
		{
			// Spread attribute: multiple spaces should be reported
			Code:   `var x = <div {...props}  foo="bar" />`,
			Tsx:    true,
			Output: []string{`var x = <div {...props} foo="bar" />`},
			Errors: []rule_tester.InvalidTestCaseError{
				{
					MessageId: "onlyOneSpace",
					Line:      1,
					Column:    26,
				},
			},
		},
		{
			// Multiple violations: tag-to-prop and prop-to-prop
			Code:   `var x = <App  foo   bar />`,
			Tsx:    true,
			Output: []string{`var x = <App foo bar />`},
			Errors: []rule_tester.InvalidTestCaseError{
				{
					MessageId: "onlyOneSpace",
					Line:      1,
					Column:    15,
				},
				{
					MessageId: "onlyOneSpace",
					Line:      1,
					Column:    21,
				},
			},
		},
		{
			// Spread with multiple spaces on both sides
			Code:   `var x = <App foo  {...test}  bar />`,
			Tsx:    true,
			Output: []string{`var x = <App foo {...test} bar />`},
			Errors: []rule_tester.InvalidTestCaseError{
				{
					MessageId: "onlyOneSpace",
					Line:      1,
					Column:    19,
				},
				{
					MessageId: "onlyOneSpace",
					Line:      1,
					Column:    30,
				},
			},
		},
		{
			// Compound component name with multiple spaces
			Code:   `var x = <Foo.Bar  baz="quux" />`,
			Tsx:    true,
			Output: []string{`var x = <Foo.Bar baz="quux" />`},
			Errors: []rule_tester.InvalidTestCaseError{
				{
					MessageId: "onlyOneSpace",
					Line:      1,
					Column:    19,
				},
			},
		},
		{
			// Multiple blank lines between props
			Code: `var x = <div
  foo="bar"

  baz="qux"

  quux="corge"
/>`,
			Tsx: true,
			Errors: []rule_tester.InvalidTestCaseError{
				{
					MessageId: "noLineGap",
					Line:      4,
					Column:    3,
				},
				{
					MessageId: "noLineGap",
					Line:      6,
					Column:    3,
				},
			},
		},
		{
			// TypeScript generic with multiple spaces
			Code:   `var x = <App<T>  foo="bar" />`,
			Tsx:    true,
			Output: []string{`var x = <App<T> foo="bar" />`},
			Errors: []rule_tester.InvalidTestCaseError{
				{
					MessageId: "onlyOneSpace",
					Line:      1,
					Column:    18,
				},
			},
		},
		{
			// Comment then blank line: only the blank line gap is reported
			Code: `var x = <button
  title="Some button"
  // this is a comment
  onClick={() => {}}

  type="button"
/>`,
			Tsx: true,
			Errors: []rule_tester.InvalidTestCaseError{
				{
					MessageId: "noLineGap",
					Line:      6,
					Column:    3,
				},
			},
		},
	})
}

File: packages/rslint-test-tools/tests/eslint-plugin-stylistic/rules/jsx-props-no-multi-spaces.test.ts
/**
 * @fileoverview Tests for jsx-props-no-multi-spaces rule.
 *
 * Ported verbatim from @stylistic/eslint-plugin v5.10.0:
 *   packages/eslint-plugin/rules/jsx-props-no-multi-spaces/jsx-props-no-multi-spaces.test.ts
 *
 * Transformations applied per the porting spec:
 *  - `run({ name, rule, parserOptions, valid, invalid })`
 *      -> `ruleTester.run('jsx-props-no-multi-spaces', null as never, { valid, invalid })`
 *  - The upstream `valids(...)` / `invalids(...)` wrappers (`#test/parsers-jsx`)
 *    are a multi-parser fan-out harness (default / `@babel/eslint-parser` /
 *    `@typescript-eslint/parser`) that also appends a `// features: [...],
 *    parser: ...` trailing comment to each fixture. That augmentation is a pure
 *    test-harness artifact, so every case is ported to its bare `code` /
 *    `output` / `errors`, evaluated to its final form.
 *  - `parserOptions` (ecmaFeatures.jsx) dropped — rslint resolves via tsconfig and
 *    the RuleTester routes JSX fixtures to `.tsx`.
 *  - Per-case `features` dropped (see below for the one that carried any).
 *  - The plain backtick multi-line templates are kept byte-for-byte (their literal
 *    leading indentation is significant and is preserved verbatim).
 *
 * Autofix: `@stylistic/jsx-props-no-multi-spaces` is fixable (meta.fixable ===
 * 'code'). The `onlyOneSpace` invalid cases pin an upstream `output` (collapsing
 * the extra spaces); they are ported with that `output`. The `noLineGap` invalid
 * cases pin NO `output` upstream (a blank line between props is reported but not
 * autofixed) — they are ported `output`-less, so the RuleTester checks the
 * diagnostics only and never asserts an invented fix.
 *
 * Feature-tagged case — verified to parse cleanly under ts-go and to match the
 * upstream expectation, so it stays in the green set:
 *  - `<App<T> foo bar />` carried `features: ['ts', 'no-babel']` (upstream runs it
 *    only on the TS parser). ts-go parses the generic JSX element without a syntax
 *    error and the rule emits 0 diagnostics, matching upstream's `valid`
 *    expectation. (Independently confirmed: the 2-space variant `<App<T>  foo bar />`
 *    reports `onlyOneSpace` with prop1 `App`, so the syntax is genuinely parsed.)
 *
 * No `._css_` / `._json_` / `._markdown_` test files exist for this rule, and the
 * single `.test.ts` has exactly one `run()` block (no skipBabel block). No
 * suggestions are pinned anywhere in the upstream file. No KNOWN GAPS surfaced:
 * every fixture parses under ts-go and aligns with upstream.
 */

import { RuleTester } from '../rule-tester';

const ruleTester = new RuleTester();

ruleTester.run('jsx-props-no-multi-spaces', null as never, {
  valid: [
    {
      code: `
        <App />
      `,
    },
    {
      code: `
        <App foo />
      `,
    },
    {
      code: `
        <App foo bar />
      `,
    },
    {
      code: `
        <App foo="with  spaces   " bar />
      `,
    },
    {
      code: `
        <App
          foo bar />
      `,
    },
    {
      code: `
        <App
          foo
          bar />
      `,
    },
    {
      code: `
        <App
          foo {...test}
          bar />
      `,
    },
    {
      code: '<App<T> foo bar />',
    },
    {
      code: '<Foo.Bar baz="quux" />',
    },
    {
      code: '<Foobar.Foo.Bar.Baz.Qux.Quux.Quuz.Corge.Grault.Garply.Waldo.Fred.Plugh xyzzy="thud" />',
    },
    {
      code: `
        <button
          title="Some button"
          type="button"
        />
      `,
    },
    {
      code: `
        <button
          title="Some button"
          onClick={(value) => {
            console.log(value);
          }}
          type="button"
        />
      `,
    },
    {
      code: `
        <button
          title="Some button"
          // this is a comment
          onClick={(value) => {
            console.log(value);
          }}
          type="button"
        />
      `,
    },
    {
      code: `
        <button
          title="Some button"
          // this is a comment
          // this is a second comment
          onClick={(value) => {
            console.log(value);
          }}
          type="button"
        />
      `,
    },
    {
      code: `
        <App
          foo="Some button" // comment
          // comment
          bar=""
        />
      `,
    },
    {
      code: `
        <button
          title="Some button"
          /* this is a multiline comment
              ...
              ... */
          onClick={(value) => {
            console.log(value);
          }}
          type="button"
        />
      `,
    },
  ],

  invalid: [
    {
      code: `
        <App  foo />
      `,
      output: `
        <App foo />
      `,
      errors: [
        {
          messageId: 'onlyOneSpace',
          data: { prop1: 'App', prop2: 'foo' },
        },
      ],
    },
    {
      code: `
        <App foo="with  spaces   "   bar />
      `,
      output: `
        <App foo="with  spaces   " bar />
      `,
      errors: [
        {
          messageId: 'onlyOneSpace',
          data: { prop1: 'foo', prop2: 'bar' },
        },
      ],
    },
    {
      code: `
        <App foo  bar />
      `,
      output: `
        <App foo bar />
      `,
      errors: [
        {
          messageId: 'onlyOneSpace',
          data: { prop1: 'foo', prop2: 'bar' },
        },
      ],
    },
    {
      code: `
        <App  foo   bar />
      `,
      output: `
        <App foo bar />
      `,
      errors: [
        {
          messageId: 'onlyOneSpace',
          data: { prop1: 'App', prop2: 'foo' },
        },
        {
          messageId: 'onlyOneSpace',
          data: { prop1: 'foo', prop2: 'bar' },
        },
      ],
    },
    {
      code: `
        <App foo  {...test}  bar />
      `,
      output: `
        <App foo {...test} bar />
      `,
      errors: [
        {
          messageId: 'onlyOneSpace',
          data: { prop1: 'foo', prop2: 'test' },
        },
        {
          messageId: 'onlyOneSpace',
          data: { prop1: 'test', prop2: 'bar' },
        },
      ],
    },
    {
      code: '<Foo.Bar  baz="quux" />',
      output: '<Foo.Bar baz="quux" />',
      errors: [
        {
          messageId: 'onlyOneSpace',
          data: { prop1: 'Foo.Bar', prop2: 'baz' },
        },
      ],
    },
    {
      code: `
        <Foobar.Foo.Bar.Baz.Qux.Quux.Quuz.Corge.Grault.Garply.Waldo.Fred.Plugh  xyzzy="thud" />
      `,
      output: `
        <Foobar.Foo.Bar.Baz.Qux.Quux.Quuz.Corge.Grault.Garply.Waldo.Fred.Plugh xyzzy="thud" />
      `,
      errors: [
        {
          messageId: 'onlyOneSpace',
          data: { prop1: 'Foobar.Foo.Bar.Baz.Qux.Quux.Quuz.Corge.Grault.Garply.Waldo.Fred.Plugh', prop2: 'xyzzy' },
        },
      ],
    },
    {
      code: `
        <button
          title='Some button'

          type="button"
        />
      `,
      errors: [
        {
          messageId: 'noLineGap',
          data: { prop1: 'title', prop2: 'type' },
        },
      ],
    },
    {
      code: `
        <button
          title="Some button"

          onClick={(value) => {
            console.log(value);
          }}

          type="button"
        />
      `,
      errors: [
        {
          messageId: 'noLineGap',
          data: { prop1: 'title', prop2: 'onClick' },
        },
        {
          messageId: 'noLineGap',
          data: { prop1: 'onClick', prop2: 'type' },
        },
      ],
    },
    {
      code: `
        <button
          title="Some button"
          // this is a comment
          onClick={(value) => {
            console.log(value);
          }}

          type="button"
        />
      `,
      errors: [
        {
          messageId: 'noLineGap',
          data: { prop1: 'onClick', prop2: 'type' },
        },
      ],
    },
    {
      code: `
        <button
          title="Some button"
          // this is a comment
          // second comment

          onClick={(value) => {
            console.log(value);
          }}

          type="button"
        />
      `,
      errors: [
        {
          messageId: 'noLineGap',
          data: { prop1: 'title', prop2: 'onClick' },
        },
        {
          messageId: 'noLineGap',
          data: { prop1: 'onClick', prop2: 'type' },
        },
      ],
    },
    {
      code: `
          <button
            title="Some button"
            /*this is a
              multiline
              comment
            */

            onClick={(value) => {
              console.log(value);
            }}

            type="button"
          />
        `,
      errors: [
        {
          messageId: 'noLineGap',
          data: { prop1: 'title', prop2: 'onClick' },
        },
        {
          messageId: 'noLineGap',
          data: { prop1: 'onClick', prop2: 'type' },
        },
      ],
    },
  ],
});
